### ISSUE TITLE
Correctly handling weight decay in the optimizer

### ISSUE THREAD

**Issue by sxjscience** [dmlc/mxnet#5099]
Our objective function is like `l(w) = E_x[f_w(x)] + \lambda ||w||_2^2`. If we include the L2 regularizer, the stochastic gradient should be `true_g = rescale_grad * g + wd * weight`. The stochastic gradient will then be used to update the curvature and momentum.

In the current implementation of Adam and RMSProp, the `g` is directly used as the stochastic gradient for updating the other state variables, which is not consistent with other packages like Keras. We may need to revise this. @dsqx71 @leezu @piiswrong @howard0su 

Reference:
-----------
Adam in MXNet: https://github.com/dmlc/mxnet/blob/master/src/operator/optimizer_op-inl.h#L196-L197
Adam in Keras: https://github.com/fchollet/keras/blob/master/keras/optimizers.py#L415-L446

**Comment by piiswrong** [comment]
yes we should change it

**Comment by leezu** [comment]
I adapted the code in the uncentered RMSProp pull request (#5116). For Adam I opened a new pull request (#5126)

**Comment by alanhuang1990** [comment]
This could be the reason of performance diff reported in the issues.  The new way to handle weight decay will cause un-convergence in my experiments.  Similar conclusion can be find in https://stats.stackexchange.com/questions/243881/weight-decay-and-rmsprop-in-neural-networks

**Comment by sxjscience** [comment]
@alanhuang1990 Are you using the non-centered version or the centered version? This is the correct way of handling weight decay and if it makes the convergence worse, you may try to tune the learning-rate again.

**Comment by alanhuang1990** [comment]
@sxjscience  Alright.  The way to handle weight decay seems to be not important in many experiments. But the way to handle 'clip_gradient' is wrong.   
https://github.com/dmlc/mxnet/blob/master/src/operator/optimizer_op-inl.h#L317

What implemented here is weight clipping,  which should not be activated in 'clip_gradient'

**Comment by sxjscience** [comment]
@alanhuang1990 It's the check that decides whether to turn on weight clipping.
```c++
if (param.clip_weights >= 0.0f) {
   Assign(out, req[0], F<clip>(weight + delta, DType(param.clip_weights)));
}
```

**Comment by sxjscience** [comment]
@alanhuang1990 Also, correcting the way we handle weight-decay is necessary for research on optimization algorithms. We need to make sure that all the optimizers are solving the same objective function.

**Comment by leezu** [comment]
@alanhuang1990 Can you clarify what you mean by 

> But the way to handle 'clip_gradient' is wrong.
> https://github.com/dmlc/mxnet/blob/master/src/operator/optimizer_op-inl.h#L317
> 
> What implemented here is weight clipping, which should not be activated in 'clip_gradient'.

The code you refer to is implementing weight clipping, which is useful for certain scenarios such as training a network according to the [WGAN paper](https://arxiv.org/abs/1701.07875). It is independent of gradient clipping.
